### PR TITLE
Add conda-forge based CI

### PR DESCRIPTION
**Pull request by jaimergp**
### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

Comes from https://github.com/conda/conda-libmamba-solver/pull/120, using https://github.com/conda/conda/pull/12335

### CODE REVIEW

**Comment by jaimergp** [comment]
Needs changes in https://github.com/conda/conda/pull/12350

**Comment by jaimergp** [comment]
@costrouc I'll need some help debugging the errors in `tests/test_create.py::IntegrationTests::test_conda_pip_interop_conda_editable_package`. Maybe we can pair-debug this one?

**Comment by costrouc** [comment]
I can reproduce this test with the following locally

```shell
docker run --rm -it \
       -v /home/costrouc/p/conda/conda:/opt/conda-src \
       -v /home/costrouc/p/conda/conda-libmamba-solver:/opt/conda-libmamba-solver-src \
       -e TEST_SPLITS=3 \
       -e TEST_GROUP=1 \
       -e CONDA_SOLVER=libmamba \
       ghcr.io/conda/conda-ci:main-linux-python3.10-conda-forge \
       bash
````

Then running 

````shell
sudo env CONDA_SOLVER=classic /opt/conda/condabin/conda install -p /opt/conda --file /opt/conda-libmamba-solver-src/dev/requirements.txt
/opt/conda/bin/python -m pip install /opt/conda-libmamba-solver-src --no-deps -vvv
source /opt/conda-src/dev/linux/integration.sh
````

I am next going to try to narrow this down.

**Comment by costrouc** [comment]
Okay can reproduce with a single test.

```shell
sudo env CONDA_SOLVER=classic /opt/conda/condabin/conda install -p /opt/conda --file /opt/conda-libmamba-solver-src/dev/requirements.txt
/opt/conda/bin/python -m pip install /opt/conda-libmamba-solver-src --no-deps -vvv

# taken from dev/linux/integration.sh
git config --global --add safe.directory /opt/conda-src
sudo su root -c "/opt/conda/bin/conda install -yq conda-build"
sudo /opt/conda/bin/conda install --quiet -y --file tests/requirements.txt
eval "$(sudo /opt/conda/bin/python -m conda init --dev bash)"
conda-build tests/test-recipes/activate_deactivate_package tests/test-recipes/pre_link_messages_package
conda info
export TMP=$HOME/pytesttmp
mkdir -p $TMP

# run single test
pytest tests/test_create.py::IntegrationTests::test_conda_pip_interop_conda_editable_package
````

**Comment by costrouc** [comment]
When using `pdb` I was able to extract some of the state around the failure.

```python
(Pdb) unlink_precs
IndexedSet([PrefixRecord(_hash=919153346818182340, name='urllib3', version='1.19.1', build='dev_0', build_number=0, channel=Channel("<develop>"), subdir='pypi', depends=('python 2.7.*',), constrains=('cryptography >=1.3.4', 'idna >=2.0.0', 'pyopenssl >=0.14', 'pysocks !=1.5.7,<2.0,>=1.5.6'), package_type='virtual_python_egg_link', files=(), paths_data=PathsData(paths_version=1, paths=()))])
(Pdb) link_precs
IndexedSet([PackageRecord(_hash=6144791784308069146, name='urllib3', version='1.20', build='py27_1', build_number=1, channel=Channel("conda-forge/linux-64"), subdir='linux-64', fn='urllib3-1.20-py27_1.tar.bz2', md5='0507cc609938ad8fc81c18f818b80f52', url='https://conda.anaconda.org/conda-forge/linux-64/urllib3-1.20-py27_1.tar.bz2', sha256='139f41180aebfbb75161e313daecc29bd34c5b914ac27948e6ba62d7ae8ab2ca', depends=('python 2.7*',), constrains=('python_abi * *_cp27mu',), track_features=(), license='MIT', timestamp=0, size=130718)])
(Pdb) unlink_precs[0]
PrefixRecord(_hash=919153346818182340, name='urllib3', version='1.19.1', build='dev_0', build_number=0, channel=Channel("<develop>"), subdir='pypi', depends=('python 2.7.*',), constrains=('cryptography >=1.3.4', 'idna >=2.0.0', 'pyopenssl >=0.14', 'pysocks !=1.5.7,<2.0,>=1.5.6'), package_type='virtual_python_egg_link', files=(), paths_data=PathsData(paths_version=1, paths=()))
(Pdb) unlink_precs[0].is_unmanageable
True
(Pdb) unlink_precs[0].package_type
<PackageType.VIRTUAL_PYTHON_EGG_LINK: 'virtual_python_egg_link'>
(Pdb) from conda.models.enums import PackageType
(Pdb) unlink_precs[0].package_type in PackageType.unmanageable_package_types()
True
(Pdb) self.specs_to_add
IndexedSet([MatchSpec("urllib3=1.20")])
(Pdb) self.prefix
'/home/test_user/pytesttmp/0b0a abdecf'
(Pdb) os.listdir(self.prefix + '/lib/python2.7/site-packages/')
['certifi-2020.6.20-py3.6.egg-info', 'README', 'urllib3.egg-link', 'wheel-0.37.1-py3.9.egg-info', 'setuptools-44.0.0.post20200106-py2.7.egg-info', 'wheel', 'pip-10.0.1-py2.7.egg-info', 'setuptools', 'easy-install.pth', 'pkg_resources', 'pip', 'certifi', 'easy_install.py', 'easy_install.pyc']
(Pdb) subprocess.check_output('ls -la "/home/test_user/pytesttmp/0b0a abdecf/lib/python2.7/site-packages/urllib3.egg-link"', shell=True)
b'-rw-r--r-- 1 test_user test_user 51 Feb 27 03:34 /home/test_user/pytesttmp/0b0a abdecf/lib/python2.7/site-packages/urllib3.egg-link\n'
````

**Comment by jaimergp** [comment]
Awesome! Thanks @costrouc. Look how fun this is getting now:

* The line that fails is 1688:

```python
# should raise an error
with pytest.raises(PackagesNotFoundError):
    # TODO: This raises PackagesNotFoundError, but the error should really explain
    #       that we can't install urllib3 because it's already installed and
    #       unmanageable. The error should suggest trying to use pip to uninstall it.
    stdout, stderr, _ = run_command(Commands.INSTALL, prefix, "urllib3=1.20", "--dry-run")  # <<< THIS ONE
```

The TODO note says that we should actually see a different exception... which is the one we are getting 😂  : 

```
________ IntegrationTests.test_conda_pip_interop_conda_editable_package ________
Traceback (most recent call last):
  File "/opt/conda-src/tests/test_create.py", line 1688, in test_conda_pip_interop_conda_editable_package
    stdout, stderr, _ = run_command(Commands.INSTALL, prefix, "urllib3=1.20", "--dry-run")
  File "/opt/conda-src/conda/testing/integration.py", line 287, in run_command
    result = do_call(args, p)
  File "/opt/conda-src/conda/cli/conda_argparse.py", line 92, in do_call
    return getattr(module, func_name)(args, parser)
  File "/opt/conda-src/conda/notices/core.py", line 121, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda-src/conda/cli/main_install.py", line 20, in execute
    install(args, parser, 'install')
  File "/opt/conda-src/conda/cli/install.py", line 264, in install
    unlink_link_transaction = solver.solve_for_transaction(
  File "/opt/conda-src/conda/core/solve.py", line 130, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
  File "/opt/conda-src/conda/core/solve.py", line 182, in solve_for_diff
    raise RuntimeError("Cannot unlink unmanageable packages:%s"
RuntimeError: Cannot unlink unmanageable packages:
  - <develop>/pypi::urllib3-1.19.1-dev_0
```


However, at no point I see the `conda_libmamba_solver` involved in that traceback. Did you set up `CONDA_SOLVER=libmamba` before running the test?

Which branches are checked out in `conda/conda` and `conda/conda-libmamba-solver`, locally?

**Comment by costrouc** [comment]
> However, at no point I see the conda_libmamba_solver involved in that traceback. Did you set up CONDA_SOLVER=libmamba before running the test?

Yeah if I look at the context around the sovler being used in the test

````
self = <conda_libmamba_solver.solver.LibMambaSolver object at 0x7f484b7dbd60>
update_modifier = <UpdateModifier.FREEZE_INSTALLED: 'freeze_installed'>
deps_modifier = <DepsModifier.NOT_SET: 'not_set'>
prune = <conda.auxlib._Null object at 0x7f486ddd3eb0>
ignore_pinned = <conda.auxlib._Null object at 0x7f486ddd3eb0>
force_remove = <conda.auxlib._Null object at 0x7f486ddd3eb0>
force_reinstall = False
should_retry_solve = True
```

> Which branches are checked out in conda/conda and conda/conda-libmamba-solver, locally?

conda https://github.com/conda/conda/commit/598bc26e4 (main)

conda-libmamba-solver https://github.com/conda/conda-libmamba-solver/commit/02423c3 (conda-forge-ci)

**Review by costrouc — APPROVED** [review]
@jaimergp looks like all test are passing and I can confirm that all packages from the `base` environment do come from `conda-forge` which was the goal of this PR. I've created an issue linked above which details the test that was disabled.

**Review by jezdez — CHANGES_REQUESTED** [review]
With https://github.com/conda/conda/pull/12297 being merged, it might be useful to adopt the same pattern to build the test matrix via the exclude statement.

**Comment by jaimergp** [comment]
We are now running _all_ conda's unit tests, which should not be needed, but we have still found some errors that need investigation. Some of them might be just that the tests are patching things only on conda, but not on conda-libmamba-solver. I'll post a summary later.

**Comment by jaimergp** [comment]
`cryptography` issues aside (see thread), this PR revealed these failing tests:

```
FAILED tests/cli/test_main_notices.py::test_notices_does_not_interrupt_command_on_failure
FAILED tests/conda_env/installers/test_pip.py::PipInstallerTest::test_stops_on_exception
FAILED tests/conda_env/installers/test_pip.py::PipInstallerTest::test_straight_install
FAILED tests/conda_env/specs/test_base.py::DetectTestCase::test_build_msg
FAILED tests/conda_env/specs/test_base.py::DetectTestCase::test_dispatches_to_registered_specs
FAILED tests/conda_env/specs/test_base.py::DetectTestCase::test_has_build_msg_function
FAILED tests/conda_env/specs/test_base.py::DetectTestCase::test_passes_kwargs_to_all_specs
FAILED tests/conda_env/specs/test_base.py::DetectTestCase::test_raises_exception_if_no_detection
FAILED tests/conda_env/specs/test_requirements.py::TestRequiremets::test_environment
ERROR tests/cli/test_main_notices.py::test_notices_appear_once_when_running_decorated_commands
FAILED tests/models/test_prefix_graph.py::test_windows_sort_orders_1
FAILED tests/models/test_prefix_graph.py::test_deep_cyclical_dependency
FAILED tests/plugins/test_solvers.py::test_get_solver_backend
FAILED tests/plugins/test_solvers.py::test_get_solver_backend_multiple
```

Some of them need to be deactivated because they target behaviour not implemented / implementable in `conda-libmamba-solver`.

**Comment by jaimergp** [comment]
We still need https://github.com/conda/conda/pull/12389

**Comment by jaimergp** [comment]
Added https://github.com/conda/conda-libmamba-solver/issues/143 to track deselected upstream tests.

**Comment by jaimergp** [comment]
Well, we won't need to worry about that py37 error anymore because we are not supporting it anyway. See https://github.com/conda/conda/issues/12436 for details.

**Comment by jaimergp** [comment]
See https://github.com/conda-forge/admin-requests/pull/687 for unrelated osx errors.

**Comment by jaimergp** [comment]
Re-running failed jobs to see if the packaging issue was fixed.

**Comment by jaimergp** [comment]
@jezdez - can you take another look? This is ready for review / merge on my end, I think.

**Review by jezdez — CHANGES_REQUESTED** [review]
I think this needs slight tweaks to the test matrix still, right?

File: .github/workflows/upstream_tests.yml [hunk 1]
```diff
@@ -1,293 +1,533 @@
...
+            test-type: 'integration'
+            test-group: '1'
+          - default-channel: 'defaults'
+            python-version: '3.9'
+            test-type: 'integration'
+            test-group: '2'
+          - default-channel: 'defaults'
+            python-version: '3.9'
+            test-type: 'integration'
+            test-group: '3'
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply .github/workflows/upstream_tests.yml:381]: Oh god it changed again in https://github.com/conda/conda/pull/12413 🤦

[the remark this answers is not in the document]
Comment (jezdez) [reply .github/workflows/upstream_tests.yml:381]: I know :( I thought you knew about this already, I'm sorry!

File: .github/workflows/upstream_tests.yml [hunk 2]
```diff
@@ -1,293 +1,533 @@
...
 
 on:
+  # NOTE: github.event context is push payload:
+  # https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#push
   push:
     branches:
       - main
-    paths-ignore:
-      - "docs/**"
+      - feature/**
```
[the remark this answers is not in the document]
Comment (jezdez) [reply .github/workflows/upstream_tests.yml:12]: Ah, d'oh, you're right!

File: .github/workflows/upstream_tests.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaimergp) [reply .github/workflows/upstream_tests.yml]: Bad timing 😂 Sync'ed again, let's see how it goes.

**Review by jezdez — APPROVED** [review]

**Pull request merged**